fx quant: preserve node stack trace throughout prepare and convert

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #70757

Summary:

This is an initial PR on a way to preserve stack traces throughout FX
graph mode quantization.  It preserves the stack traces for ops
for all of the quantize handlers. A future PR will add stack traces
for dtype transitions.

Test Plan:

```
python test/test_quantization.py
TestQuantizeFx.test_stack_trace_preserved
```

Note: the above only tests a single case. In a future PR, once we
expand coverage, we can expand the utility functions to check for stack
traces on all tests.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33432485](https://our.internmc.facebook.com/intern/diff/D33432485)

vkuzo committed:
[wip] preserve node stack trace throughout FX graph mode quantization

Summary:

This is an initial PR on a way to preserve stack traces throughout FX
graph mode quantization.  It contains an example on how this could work,
implemented for only one callsite (torch.add + int8 quant).

We should go through initial review before moving forward.

Test Plan:

```
python test/test_quantization.py
TestQuantizeFx.test_stack_trace_preserved
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33432485).

jerryzh168 commented on torch/ao/quantization/fx/quantization_patterns.py:
```diff
@@ -459,8 +460,11 @@ def modified_load_arg(n: Node):
-                    op = quantized_graph.create_node(
-                        'call_function', self.quantized_binary_op, add_args, kwargs)
+                    # TODO(after initial review): expand to all callsites in this file
+                    op = create_node_from_old_node_preserve_meta(
```
can we add to the reference path? this path is going to be deprecated, the ops that's using reference path can be found in https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/fx/_lower_to_native_backend.py and  https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/fx/quantized_fusion_patterns_and_replacements.py

vkuzo replied:
sure.  The current form of this PR is just to align on how this will look.  If we think this way is the way to go, we can add it to all callsites.

kimishpatel replied:
I dont know the details of FX quantization but the fusion passes (in this example add + relu) will also work? and if so which node's stack trace will be copied.

vkuzo replied:
> I dont know the details of FX quantization but the fusion passes (in this example add + relu) will also work? and if so which node's stack trace will be copied.

the simplest thing to do is only preserve stack traces when the new node is created directly from a single old node, which is what this PR implements.  We can discuss what to do about fusions and nodes which do not appear in the original graph such as quant/dequant in a future PR.

jerryzh168 reviewed (approved):
Looks good I think, let's verify the referene path works and only add this to reference path, since the current non-reference branch in quantization_patterns.py will be deprecated in the future

jerryzh168 commented:
> > Looks good I think, let's verify the referene path works and only add this to reference path, since the current non-reference branch in quantization_patterns.py will be deprecated in the future
> 
> do we have context on the deprecation, the timeline, how to enable reference model path, etc? The test case hits the path which I modified, so to the user it seems like the reference model path is not being hit unless they know how to enable it.

oh, the deprecation is going on right now and probably take a few months. It's not a user visible change, user don't need to do anything to enable the path.

vkuzo committed:
Update on "[wip] preserve node stack trace throughout FX graph mode quantization"


Summary:

This is an initial PR on a way to preserve stack traces throughout FX
graph mode quantization.  It contains an example on how this could work,
implemented for only one callsite (torch.add + int8 quant).

We should go through initial review before moving forward.

Test Plan:

```
python test/test_quantization.py
TestQuantizeFx.test_stack_trace_preserved
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33432485](https://our.internmc.facebook.com/intern/diff/D33432485)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33432485).

vkuzo commented:
ok, I changed the test case to Linear which is already supported in reference path, please take a look again.

vkuzo committed:
Update on "fx quant: preserve node stack trace throughout prepare and convert"


Summary:

This is an initial PR on a way to preserve stack traces throughout FX
graph mode quantization.  It preserves the stack traces for ops
for all of the quantize handlers. A future PR will add stack traces
for dtype transitions.

Test Plan:

```
python test/test_quantization.py
TestQuantizeFx.test_stack_trace_preserved
```

Note: the above only tests a single case. In a future PR, once we
expand coverage, we can expand the utility functions to check for stack
traces on all tests.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33432485](https://our.internmc.facebook.com/intern/diff/D33432485)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33432485).

vkuzo committed:
Update on "fx quant: preserve node stack trace throughout prepare and convert"


Summary:

This is an initial PR on a way to preserve stack traces throughout FX
graph mode quantization.  It preserves the stack traces for ops
for all of the quantize handlers. A future PR will add stack traces
for dtype transitions.

Test Plan:

```
python test/test_quantization.py
TestQuantizeFx.test_stack_trace_preserved
```

Note: the above only tests a single case. In a future PR, once we
expand coverage, we can expand the utility functions to check for stack
traces on all tests.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33432485](https://our.internmc.facebook.com/intern/diff/D33432485)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33432485).

jerryzh168 commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -3511,6 +3511,44 @@ def forward(self, x):
+        class M(nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.linear = nn.Linear(1, 1)
```
could we check if the stack trace is preserved for lowering, we can do that by testing quantization for a single functional relu

vkuzo replied:
sure, planning to do in a future PR, hope that's ok
